fbsat-core/src/main/kotlin/ru/ifmo/fbsat/core/task/single/complete/Cegis.kt

```diff
@@ -129,6 +127,38 @@ fun Inferrer.performCegis(smvDir: File, loopNumber: Int): Automaton? {
             )
             return null
         }
+
+        val maxLabel = if (Globals.NEGATIVE_TREE_OPTIMIZATIONS == NegativeTreeOptimizations.OPT1) {
+            val prevMaxLabel = negativeScenarioTree.nodes.maxOfOrNull { it.label }
+            if (prevMaxLabel != null) {
+                val visitedNodes = mutableSetOf<NegativeScenarioTree.Node>()
+                val visitedFullScenarios = negativeScenarioTree.scenarios.filter { scenario ->
+                    val path = automaton.eval(scenario).toList()
+                    for ((scenarioElement, pathElement) in scenario.elements zip path) {
+                        val node = negativeScenarioTree.nodes[scenarioElement.nodeId!! - 1]
+                        check(scenarioElement.nodeId == node.id) { "Non-equal node ids" }
+                        visitedNodes += node
+                        if (scenarioElement.outputAction != pathElement.outputAction) {
+                            return@filter false
+                        }
+                    }
+                    scenario.loopPosition == null || path[scenario.loopPosition - 1].destination == path.last().destination
+                }
+                val maxLabel = prevMaxLabel + 1
+                for (node in visitedNodes) {
+                    node.label = maxLabel
+                }
+                if (visitedFullScenarios.isNotEmpty()) {
+                    logger.warn("Visited full negative-scenarios:\n------------\n${(visitedFullScenarios zip visitedFullScenarios.map { automaton.eval(it).toList() }).joinToString("\n\t\t") {
```

Comment:
I bet this has to be just "debug". Make sure that we don't have warn-spam when running without --debug flag.